### PR TITLE
feat: add user notification preferences with configurable email, SMS, and push notification

### PR DESCRIPTION
**Pull request by Sarthak-btr**
**User Notification Preferences**

This PR adds comprehensive user notification preferences functionality, allowing users to configure their communication preferences (email, SMS, and push notifications) on a per-account basis. The implementation follows a clean architecture pattern with proper separation of concerns across repository, service, and API layers.

---

## Key Features

* **Configurable Preferences**: Users can independently enable/disable email, SMS, and push notifications
* **Service Integration**: Notification services check user preferences before sending communications
* **Default Settings**: All notification channels enabled by default for new users
* **Preference Override**: Administrative/critical notifications can bypass user preferences when needed
* **Secure Access**: Authentication and authorization controls ensure users can only modify their own preferences
* **Efficient Data Model**: Optimized MongoDB schema with proper indexing and validation

---

### Get Account with Notification Preferences

**GET** `/accounts/<id>?include_notification_preferences=true`

**Response**: Standard account object with optional notification preferences

```json
{
  "id": "string",
  "username": "string",
  "first_name": "string",
  "last_name": "string",
  "notification_preferences": {
    "account_id": "string",
    "email_enabled": true,
    "push_enabled": true,
    "sms_enabled": true
  }
}
```

### Update Notification Preferences

**PUT** `/accounts/<account_id>/notification-preferences`

**Request Body**:

```json
{
  "email_enabled": boolean,
  "push_enabled": boolean,
  "sms_enabled": boolean
}
```

**Response**: Updated preferences object

```json
{
  "account_id": "string",
  "email_enabled": boolean,
  "push_enabled": boolean,
  "sms_enabled": boolean
}
```

---

## Database Schema Changes

Added a new collection `account_notification_preferences` with the following schema:

```json
{
  "_id": ObjectId,
  "account_id": String,
  "email_enabled": Boolean,
  "push_enabled": Boolean,
  "sms_enabled": Boolean,
  "active": Boolean,
  "created_at": Date,
  "updated_at": Date
}
```

**Indexes:**

* Unique compound index on `(active, account_id)` with partial filter for active records
* Non-unique index on `account_id` for efficient lookups

---

## Service Integration

Modified existing notification services to respect user preferences:

* **Email Service**: Updated to check preferences before sending
* **SMS Service**: Updated to check preferences before sending
* **Future-ready**: Architecture supports adding push notifications with minimal changes

---

## Tests

**Automated Tests**

* **Unit Tests:**

  * `test_notification_preferences_service.py`: Tests service-layer functionality
  * `test_notification_preferences_api.py`: Tests API endpoints and error handling

**Test Coverage:**

* Creation of default preferences
* Reading preferences
* Updating preferences (creating when non-existent)
* Authentication and authorization checks
* Edge cases (all notifications disabled, etc.)
* Integration with notification services

#### Manual Test Cases

1. **Default Preferences Verification**

   * Create a new user via `POST /accounts`
   * Call `GET /accounts/{id}?include_notification_preferences=true`
   * Verify all preferences default to `true`

2. **Update Notification Preferences**

   * Authenticate as the user
   * Call `PUT /accounts/{account_id}/notification-preferences` with:

     ```json
     {"email_enabled": false, "sms_enabled": false}
     ```
   * Verify response shows updated preferences
   * Call `GET` endpoint to confirm persistence

3. **Notification Service Integration**

   * Disable email notifications for a user
   * Trigger an action that sends an email
   * Verify email is skipped and log message is generated
   * Test override using `bypass_preferences=true`

---

### Security Testing

* Attempt to update another user's preferences without proper auth → expect **401 Unauthorized**
* Test with expired/invalid tokens → verify rejection and error messages

---

### CODE REVIEW

File: src/apps/backend/modules/account/account_service.py [hunk 1]
```diff
@@ -63,6 +68,21 @@ def get_account_by_username(*, username: str) -> Account:
     def get_account_by_username_and_password(*, params: AccountSearchParams) -> Account:
         return AccountReader.get_account_by_username_and_password(params=params)
 
+    @staticmethod
+    def get_notification_preferences(account_id: str) -> NotificationPreferences:
+        return AccountNotificationPreferenceReader.get_notification_preferences_by_account_id(account_id)
+
+    @staticmethod
+    def update_notification_preferences(params: UpdateNotificationPreferencesParams) -> NotificationPreferences:
```
Comment (jjalan) [inline src/apps/backend/modules/account/account_service.py:76]: 1. Function name is misleading - it should probably be `create_or_update_notification_preferences`. Update test name and other places based on this update.
2. Pull `account_id` from `UpdateNotificationPreferencesParams` and as an argument into `update_notification_preferences`

File: src/apps/backend/modules/account/internal/account_notification_preference_reader.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    AccountNotificationPreferencesRepository,
+)
+from modules.notification.types import NotificationPreferences
+
+
+class AccountNotificationPreferenceReader:
+    @staticmethod
+    def get_notification_preferences_by_account_id(account_id: str) -> NotificationPreferences:
+        notification_preferences = AccountNotificationPreferencesRepository.collection().find_one(
+            {"account_id": account_id}
```
Comment (jjalan) [inline src/apps/backend/modules/account/internal/account_notification_preference_reader.py:13]: Please make sure `account_id` is indexed on `AccountNotificationPreferencesRepository`, if not already, to make this query efficient.

File: src/apps/backend/modules/account/internal/account_notification_preference_reader.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        )
+
+        if notification_preferences is None:
+            default_preferences = AccountNotificationPreferencesModel(account_id=account_id, id=None).to_bson()
+            AccountNotificationPreferencesRepository.collection().insert_one(default_preferences)
+            return NotificationPreferences()
```
Comment (jjalan) [inline src/apps/backend/modules/account/internal/account_notification_preference_reader.py:19]: Can you see how other places, after insert or update, the object is returned?

File: src/apps/backend/modules/account/internal/account_notification_preference_reader.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+        preferences_model = AccountNotificationPreferencesModel.from_bson(notification_preferences)
+        return NotificationPreferences(
```
Comment (jjalan) [inline src/apps/backend/modules/account/internal/account_notification_preference_reader.py:22]: Can you see how other places, after insert or update, the object is returned?

File: src/apps/backend/modules/account/internal/account_notification_preference_reader.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            email_enabled=preferences_model.email_enabled,
+            push_enabled=preferences_model.push_enabled,
+            sms_enabled=preferences_model.sms_enabled,
+        )
+
+    @staticmethod
+    def get_existing_notification_preferences_by_account_id(account_id: str) -> dict[str, Any] | None:
```
Comment (jjalan) [inline src/apps/backend/modules/account/internal/account_notification_preference_reader.py:29]: The expected return type is `Optional[NotificationPreferences]`. Please update implementation accordingly.

**Review by jjalan — COMMENTED** [review]
I think we need to follow the code architecture closely on how its laid out. I reviewed first few files to give an idea where you differ. There might be other places that this still exists.

File: src/apps/backend/modules/account/account_service.py [hunk 6]
```diff
@@ -12,13 +14,26 @@
...
 class AccountService:
     @staticmethod
     def create_account_by_username_and_password(*, params: CreateAccountByUsernameAndPasswordParams) -> Account:
         return AccountWriter.create_account_by_username_and_password(params=params)
 
+    @staticmethod
+    def create_or_update_notification_preferences(
+        *, account_id: str, preferences: NotificationPreferencesParams
+    ) -> NotificationPreferencesParams:
+        existing_preferences = AccountNotificationPreferenceReader.get_existing_notification_preferences_by_account_id(
```
Comment (jjalan) [inline src/apps/backend/modules/account/account_service.py:29]: Create a method `create_or_update_notification_preferences` in `AccountNotificationPreferenceWriter` so the implementation in service class is simply a proxy

File: src/apps/backend/modules/account/rest_api/account_router.py [hunk 7]
```diff
@@ -9,4 +10,11 @@ def create_route(*, blueprint: Blueprint) -> Blueprint:
         blueprint.add_url_rule("/accounts", view_func=AccountView.as_view("account_view"))
         blueprint.add_url_rule("/accounts/<id>", view_func=AccountView.as_view("account_view_by_id"), methods=["GET"])
         blueprint.add_url_rule("/accounts/<id>", view_func=AccountView.as_view("account_update"), methods=["PATCH"])
+
+        blueprint.add_url_rule(
```
Comment (jjalan) [inline src/apps/backend/modules/account/rest_api/account_router.py:14]: Do not use `NotificationPreferencesView` in `AccountRouter` as it breaks module boundaries.

You can continue to call a method on `AccountService` to create_or_update_notification_preference, and in turn that can call NotificationService.create_or_update_notification_preference_for_account (or something similar).

File: src/apps/backend/modules/account/account_service.py [hunk 8]
```diff
@@ -66,3 +68,15 @@ def get_account_by_username_and_password(*, params: AccountSearchParams) -> Acco
     @staticmethod
     def update_account_profile(*, account_id: str, params: UpdateAccountProfileParams) -> Account:
         return AccountWriter.update_account_profile(account_id=account_id, params=params)
+
+    @staticmethod
+    def create_or_update_account_notification_preferences(
+        *, account_id: str, preferences: NotificationPreferencesParams
+    ) -> NotificationPreferencesParams:
```
Comment (jjalan) [inline src/apps/backend/modules/account/account_service.py:75]: nit: `NotificationPreferencesParams` - Why do we have `Params` at the end? It may make sense when bundling add, edit params in a dataclass, but when thinking OOPs, its not a good candidate for model name.

It also seems odd that input and output type is same.

File: src/apps/backend/modules/notification/email_service.py [hunk 9]
```diff
@@ -1,8 +1,30 @@
...
+from modules.logger.logger import Logger
 from modules.notification.internals.sendgrid_service import SendGridService
+from modules.notification.internals.account_notification_preferences_reader import AccountNotificationPreferenceReader
 from modules.notification.types import SendEmailParams
 
 
 class EmailService:
     @staticmethod
-    def send_email(*, params: SendEmailParams) -> None:
+    def send_email(
```
Comment (jjalan) [inline src/apps/backend/modules/notification/email_service.py:11]: rename `send_email_for_account`

File: src/apps/backend/modules/account/rest_api/account_view.py [hunk 10]
```diff
@@ -38,6 +40,20 @@ def get(self, id: str) -> ResponseReturnValue:
...
+
+        include_notification_preferences = request.args.get("include_notification_preferences", "").lower() == "true"
+
+        if include_notification_preferences:
+            try:
+                notification_preferences = AccountService.get_account_notification_preferences_by_account_id(
+                    account_id=account.id
+                )
+                preferences_dict = asdict(notification_preferences)
+                preferences_dict.pop("account_id", None)
```
Comment (jjalan) [inline src/apps/backend/modules/account/rest_api/account_view.py:52]: Why is this needed?

File: src/apps/backend/modules/account/rest_api/account_view.py [hunk 11]
```diff
@@ -58,3 +74,29 @@ def patch(self, id: str) -> ResponseReturnValue:
...
+            push_enabled=request_data.get("push_enabled", True),
+            sms_enabled=request_data.get("sms_enabled", True),
+        )
+
+        updated_preferences = AccountService.create_or_update_account_notification_preferences(
+            account_id=account_id, preferences=preferences_params
+        )
+
+        preferences_dict = asdict(updated_preferences)
+        preferences_dict.pop("account_id", None)
```
Comment (jjalan) [inline src/apps/backend/modules/account/rest_api/account_view.py:100]: Why is this needed?

File: src/apps/backend/modules/notification/email_service.py [hunk 12]
```diff
@@ -1,8 +1,28 @@
...
 from modules.notification.internals.sendgrid_service import SendGridService
+from modules.notification.internals.account_notification_preferences_reader import AccountNotificationPreferenceReader
 from modules.notification.types import SendEmailParams
 
 
 class EmailService:
     @staticmethod
-    def send_email(*, params: SendEmailParams) -> None:
+    def send_email_for_account(
+        *, account_id: Optional[str] = None, bypass_preferences: bool = False, params: SendEmailParams
```
Comment (jjalan) [inline src/apps/backend/modules/notification/email_service.py:12]: If method name is `send_email_for_account`, it is expected for `account_id` to be present.

File: src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+        return AccountNotificationPreferenceUtil.convert_account_notification_preferences_bson_to_account_params(
+            created_preferences
+        )
+
+    @staticmethod
+    def create_or_update_account_notification_preferences(
+        account_id: str, preferences: CreateOrUpdateAccountNotificationPreferences
+    ) -> AccountNotificationPreferences:
+        from modules.notification.internals.account_notification_preferences_reader import (
```
Comment (jjalan) [inline src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py:42]: If you notice the codebase, we do not do inline imports.

Please remove this.

File: src/apps/backend/modules/notification/sms_service.py [hunk 14]
```diff
@@ -1,15 +1,34 @@
...
 from modules.logger.logger import Logger
 from modules.notification.internals.twilio_service import TwilioService
+from modules.notification.internals.account_notification_preferences_reader import AccountNotificationPreferenceReader
 from modules.notification.types import SendSMSParams
 
 
 class SMSService:
     @staticmethod
-    def send_sms(*, params: SendSMSParams) -> None:
+    def send_sms_for_account(
```
Comment (jjalan) [inline src/apps/backend/modules/notification/sms_service.py:12]: If method name is send_sms_for_account, it is expected for account_id to be present.

File: src/apps/backend/modules/notification/types.py [hunk 15]
```diff
@@ -15,6 +15,21 @@ class EmailRecipient:
     email: str
 
 
+@dataclass(frozen=True)
+class CreateOrUpdateAccountNotificationPreferences:
```
Comment (sn1f3rt) [inline src/apps/backend/modules/notification/types.py:19]: This has to have `Params` at the end - `CreateOrUpdateAccountNotificationPreferencesParams`

File: src/apps/backend/modules/account/rest_api/account_view.py [hunk 16]
```diff
@@ -58,3 +72,26 @@ def patch(self, id: str) -> ResponseReturnValue:
...
+        request_data = request.get_json()
+
+        if not request_data:
+            raise AccountBadRequestError("Request body is required")
+
+        for field in ["email_enabled", "push_enabled", "sms_enabled"]:
+            if field in request_data and not isinstance(request_data[field], bool):
+                raise ValidationError(f"{field} must be a boolean")
+
+        preferences_params = CreateOrUpdateAccountNotificationPreferences(
```
Comment (sn1f3rt) [inline src/apps/backend/modules/account/rest_api/account_view.py:87]: This logic is counter-intuitive, just because the preference wasn't passed doesn't mean we should assume it is True. Instead we should selectively update only those preferences which were passed

File: src/apps/backend/modules/account/rest_api/account_router.py [hunk 17]
```diff
@@ -9,4 +9,11 @@ def create_route(*, blueprint: Blueprint) -> Blueprint:
...
+            "/accounts/<account_id>/notification-preferences",
+            view_func=AccountView.as_view("account_notification_preferences_update"),
+            methods=["PUT"],
```
Comment (sn1f3rt) [inline src/apps/backend/modules/account/rest_api/account_router.py:16]: should be patch

File: src/apps/backend/modules/notification/internals/store/account_notification_preferences_model.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+from bson import ObjectId
+
+from modules.application.base_model import BaseModel
+
+
+@dataclass
+class AccountNotificationPreferencesModel(BaseModel):
+    account_id: str
+    id: Optional[ObjectId | str]
```
Comment (sn1f3rt) [inline src/apps/backend/modules/notification/internals/store/account_notification_preferences_model.py:13]: should be set = None

File: src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    def create_account_notification_preferences(
+        account_id: str, preferences: CreateOrUpdateAccountNotificationPreferences
+    ) -> AccountNotificationPreferences:
+        preferences_model = AccountNotificationPreferencesModel(
+            account_id=account_id,
+            id=None,
+            email_enabled=preferences.email_enabled,
+            push_enabled=preferences.push_enabled,
+            sms_enabled=preferences.sms_enabled,
+            active=True,
```
Comment (sn1f3rt) [inline src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py:27]: We shouldn't pass active, id, created_at and updated_at here, they are already assigned default values in the model

**Review by sn1f3rt — CHANGES_REQUESTED** [review]

File: src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+from modules.notification.errors import AccountNotificationPreferencesNotFoundError
+from modules.notification.types import (
+    CreateOrUpdateAccountNotificationPreferencesParams,
+    AccountNotificationPreferences,
+)
+
+
+class AccountNotificationPreferenceWriter:
+    @staticmethod
+    def create_account_notification_preferences(
```
Comment (sn1f3rt) [inline src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py:22]: nit: should be private method since it is only to be used internally within this class -> `_create_account_notification_preferences`

File: src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+        account_id: str, preferences: CreateOrUpdateAccountNotificationPreferencesParams
+    ) -> AccountNotificationPreferences:
+        try:
+            AccountNotificationPreferenceReader.get_account_notification_preferences_by_account_id(account_id)
+            return AccountNotificationPreferenceWriter.update_account_notification_preferences(account_id, preferences)
+        except AccountNotificationPreferencesNotFoundError:
+            return AccountNotificationPreferenceWriter.create_account_notification_preferences(account_id, preferences)
+
+    @staticmethod
+    def update_account_notification_preferences(
```
Comment (sn1f3rt) [inline src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py:51]: nit: should be private method since it is only to be used internally within this class -> `_update_account_notification_preferences`

File: src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+
+        query = AccountNotificationPreferencesRepository.collection().insert_one(preferences_model)
+        created_preferences = AccountNotificationPreferencesRepository.collection().find_one({"_id": query.inserted_id})
+
+        return AccountNotificationPreferenceUtil.convert_account_notification_preferences_bson_to_account_notification_preferences(
+            created_preferences
+        )
+
+    @staticmethod
+    def create_or_update_account_notification_preferences(
```
Comment (sn1f3rt) [inline src/apps/backend/modules/notification/internals/account_notification_preferences_writer.py:41]: nit: let's reorder so that the private methods come first, and then the methods which are being exposed by the writer

**Review by sn1f3rt — APPROVED** [review]
Good to merge after addressing the nits.

File: src/apps/backend/modules/account/account_service.py [hunk 23]
```diff
@@ -66,3 +68,15 @@ def get_account_by_username_and_password(*, params: AccountSearchParams) -> Acco
     @staticmethod
     def update_account_profile(*, account_id: str, params: UpdateAccountProfileParams) -> Account:
         return AccountWriter.update_account_profile(account_id=account_id, params=params)
+
+    @staticmethod
+    def create_or_update_account_notification_preferences(
```
[the remark this answers is not in the document]
Comment (sn1f3rt) [reply src/apps/backend/modules/account/account_service.py:73]: @Sarthak-btr @jjalan ... this is still unclear to me
